File: android/src/main/java/com/eugeniuszx/reactnativemediapicker/OutputPlan.kt
package com.eugeniuszx.reactnativemediapicker

internal enum class RequestedFormat {
  ORIGINAL,
  JPEG,
  PNG,
  ;

  companion object {
    fun from(raw: String?): RequestedFormat = when (raw) {
      "jpeg" -> JPEG
      "png" -> PNG
      else -> ORIGINAL
    }
  }
}

internal data class OutputPlan(
  val forceReencode: Boolean,
  val preserveAnimation: Boolean,
  val target: MediaFormat.OutputFormat,
) {
  companion object {
    fun resolve(
      sourceMime: String,
      requested: RequestedFormat,
      isAnimatedSource: Boolean,
    ): OutputPlan = when (requested) {
      RequestedFormat.ORIGINAL -> OutputPlan(
        forceReencode = false,
        preserveAnimation = isAnimatedSource,
        target = MediaFormat.reencodeFormat(sourceMime),
      )
      RequestedFormat.JPEG -> explicit(sourceMime, MediaFormat.OutputFormat.JPEG)
      RequestedFormat.PNG -> explicit(sourceMime, MediaFormat.OutputFormat.PNG)
    }

    private fun explicit(
      sourceMime: String,
      target: MediaFormat.OutputFormat,
    ): OutputPlan = OutputPlan(
      forceReencode = sourceMime != MediaFormat.reencodeMime(target),
      preserveAnimation = false,
      target = target,
    )
  }
}

File: ios/Core/OutputPlan.swift
import Foundation

struct OutputPlan: Equatable {
  let forceReencode: Bool
  let preserveAnimation: Bool
  let target: ImageFormat

  static func resolve(
    source: ImageFormat,
    requested: RequestedFormat,
    isAnimatedSource: Bool
  ) -> OutputPlan {
    switch requested {
    case .original:
      return OutputPlan(
        forceReencode: false,
        preserveAnimation: isAnimatedSource,
        target: source.reencodeFormat
      )
    case .jpeg:
      return explicit(source: source, target: .jpeg)
    case .png:
      return explicit(source: source, target: .png)
    }
  }

  private static func explicit(source: ImageFormat, target: ImageFormat) -> OutputPlan {
    OutputPlan(
      forceReencode: source != target,
      preserveAnimation: false,
      target: target
    )
  }
}
